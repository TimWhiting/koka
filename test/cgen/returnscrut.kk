// Issue: https://github.com/koka-lang/koka/issues/644
fun main()
  println("Compiled")

type token
  Int

value struct token-iter<h>
  chars: vector<char>
  idx: ref<h, int>

fun token-iter/next(token-iter: token-iter<h>): <div, read<h>, write<h>> maybe<(int, token)>
  // Strange code below to work around issues with early returns in Koka 3.1.2.
  // (https://github.com/koka-lang/koka/discussions/643)

  // Skip whitespace
  val next-char0 =
    match Nothing
      Just((idx, c)) ->
        if True then
          return token-iter.next
        else
          Nothing
      Nothing -> Nothing
  val tmp = match next-char0
          Nothing -> return Nothing
          Just(c) -> c
  val (char-idx, char) = tmp
  Nothing