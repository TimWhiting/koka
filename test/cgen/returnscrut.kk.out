Compiled
test/cgen/returnscrut@kk(18, 7): type warning: pattern variable c is unused (or a wrongly spelled constructor?)
 hint: prepend an underscore to make it a wildcard pattern
test/cgen/returnscrut@kk(27, 7): type warning: pattern variable char is unused (or a wrongly spelled constructor?)
 hint: prepend an underscore to make it a wildcard pattern